vault/resource_kmip_secret_scope.go

```diff
@@ -0,0 +1,136 @@
+		Read:   kmipSecretScopeRead,
+		Update: kmipSecretScopeUpdate,
+		Delete: kmipSecretScopeDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"path": {
+				Type:         schema.TypeString,
+				Required:     true,
+				Description:  "Path where KMIP backend is mounted",
+				ValidateFunc: validateNoTrailingLeadingSlashes,
+			},
+			"scope": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Name of the scope",
+			},
+			"force": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     false,
+				Description: "Force deletion even if there are managed objects in the scope",
+			},
+		},
+	}
+}
+
+func kmipSecretScopeCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	path := d.Get("path").(string)
+	scope := d.Get("scope").(string)
+	force := d.Get("force").(bool)
+
+	data := map[string]interface{}{
+		"scope": scope,
+		"force": force,
+	}
+
+	scopePath := path + "/scope/" + scope
+	log.Printf("[DEBUG] Updating %q", scopePath)
+	if _, err := client.Logical().Write(scopePath, data); err != nil {
+		return fmt.Errorf("error updating KMIP scope %s: %s", scopePath, err)
+	}
+	d.SetId(scopePath)
+	if err := d.Set("scope", scope); err != nil {
+		return err
+	}
+
+	return kmipSecretScopeRead(d, meta)
+}
+
+func kmipSecretScopeRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	path := d.Get("path").(string)
+	expectedScope := d.Get("scope").(string)
+
+	log.Printf("[DEBUG] Reading KMIP scope at %s", path+"/scope")
+	err := readScopeFromScopeList(client, path+"/scope", expectedScope)
+	if err != nil {
+		log.Printf("[WARN] KMIP scopes not found, removing from state")
+		d.SetId("")
+		return err
+	}
+
+	return nil
+}
+
+func kmipSecretScopeUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	scope := d.Get("scope").(string)
+
+	if d.HasChange("path") {
+		newMountPath := d.Get("path").(string)
+		log.Printf("[DEBUG] Confirming KMIP scope exists at %s", newMountPath+"/scope")
+		err := readScopeFromScopeList(client, newMountPath+"/scope", scope)
+		if err != nil {
+			return fmt.Errorf("error remounting KMIP scope to new backend path %s, err=%w", newMountPath+"/scope", err)
+		}
+		d.SetId(newMountPath + "/scope/" + scope)
+	}
+
+	return kmipSecretScopeRead(d, meta)
+}
+
+func kmipSecretScopeDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	scopePath := d.Id()
+
+	log.Printf("[DEBUG] Deleting KMIP scope %q", scopePath)
+	_, err := client.Logical().Delete(scopePath)
+	if err != nil {
+		return fmt.Errorf("error deleting scope %q", scopePath)
+	}
+	log.Printf("[DEBUG] Deleted KMIP scope %q", scopePath)
+
+	return nil
+}
+
+func readScopeFromScopeList(client *api.Client, scopePath, expectedScope string) error {
```

Comment:
I feel like this function should be renamed to something  like `isScopeConfigured()` or `isScopeInPath()`
```suggestion
func isScopeConfigured(client *api.Client, path, name string) (bool, error) {
```

Reply:
What would be the purpose of returning the `bool`? Can't we use `error == nil ?` to determine if the resource is configured or not? I am under the assumption that if we do not find the scope in the scope list, i.e., it is not configured, then that is an error. 

'We should only be in a Read function after a Create or Update is called, so the resource should always exist unless there is an error' — is this a right statement?

Reply:
Added a refactor in ae7ac1f, lmk if this is along the lines of what you had in mind

Reply:
Fixed wrt feedback in c0cd973